x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/TypeResolutionUtils.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.type.DataTypes;
+
+import java.util.Locale;
+import java.util.StringJoiner;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.sql.expression.Expression.TypeResolution;
+import static org.elasticsearch.xpack.sql.expression.Expressions.ParamOrdinal;
+import static org.elasticsearch.xpack.sql.expression.Expressions.name;
+import static org.elasticsearch.xpack.sql.type.DataType.BOOLEAN;
+
+public final class TypeResolutionUtils {
```

Comment:
I would prefer using the plural instead of Utils - `TypeResolutions`. Utils has been used where the plural already existed in the rest of the codebase to avoid name clashes.

Reply:
The plural just sounded to me a bit weird, but I'll change.